Build Windows compatible font

RoelN opened the pull request:
This PR will have `make` output two fonts:

`NotoColorEmoji.ttf`, the unchanged regular font.

`NotoColorEmoji_WindowsCompatible.ttf`, which is the regular font, plus a `glyf` table with empty glyphs for each emoji, and a cmap format 4 3/1 table. This is needed by Windows 10 to recognise the font as valid font. This adds about ~8KB to the font.

This fixes https://github.com/googlefonts/noto-emoji/issues/43.

Note that the regular `NotoColorEmoji.ttf` is slightly bigger then the one [currently in the fonts directory](https://github.com/googlefonts/noto-emoji/blob/master/fonts/NotoColorEmoji.ttf). This appears to be because of how the PNG images are compressed. I'm using the tooling and instructions from the repo.

anthrotype commented on add_glyphs.py:
```diff
@@ -158,10 +161,17 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
+  if 'glyf' in font:
+    pen = TTGlyphPen(None)
+    empty_glyph = pen.glyph()
+    glyf = font['glyf']
```
I would add an `else: glyf = None` after this, then later when you add the empty glyphs you check `if glyf is not None:`.

RoelN replied:
Good call, thanks!

anthrotype commented on add_glyphs.py:
```diff
@@ -327,14 +339,33 @@ def add_ligature(lookup, cmap, seq, name):
+  # Format 4 only has unicode values 0x0000 to 0xFFFF
+  for cp, name in cmap.items():
+    if cp < 65535:
+      newtable.cmap[cp] = name
```
you could make this in one line with a dict comprehension

```python
newtable.cmap = {cp: name for cp, name in cmap.items() if cp <= 0xFFFF)
```

anthrotype commented on add_glyphs.py:
```diff
@@ -327,14 +339,33 @@ def add_ligature(lookup, cmap, seq, name):
-def update_font_data(font, seq_to_advance, vadvance, aliases):
+  # Format 4 only has unicode values 0x0000 to 0xFFFF
+  for cp, name in cmap.items():
+    if cp < 65535:
```
Should this be `<= 65535`?

RoelN replied:
Well spotted!

anthrotype reviewed (commented):
Thanks Roel. I left some comments below.

What's in "NotoColorEmoji_WindowsCompatible.tmpl.ttx.tmpl"? How does it differ from the other template ttx? Can you reuse the same template and add additional data via code?

RoelN commented:
@anthrotype Thanks for the review! I addressed your comments. In regards to "Can you reuse the same template and add additional data via code?", I'm now re-using the original template and adding the `glyf` and `loca` during build, similar to how the new `cmap` is added. This is indeed cleaner, thanks for the pointer.

anthrotype commented on add_glyphs.py:
```diff
@@ -163,6 +167,16 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
+    font['loca'] = _l_o_c_a.table__l_o_c_a()
+    font['glyf'] = glyf_table = _g_l_y_f.table__g_l_y_f()
+    glyf_table.glyphOrder = font.getGlyphOrder()
+    glyf_table.glyphs = {g:empty_glyph for g in glyf_table.glyphOrder}
```
```suggestion
    glyf_table.glyphs = {g: empty_glyph for g in glyf_table.glyphOrder}
```

anthrotype commented on add_glyphs.py:
```diff
@@ -163,6 +167,16 @@ def add_glyph_data(font, seqs, seq_to_advance, vadvance):
+  if add_glyf:
+    pen = TTGlyphPen(None)
+    empty_glyph = pen.glyph()
+    font['loca'] = _l_o_c_a.table__l_o_c_a()
```
you could also use the newTable function from fontTools.ttLib, which returns a new empty instance of table with given tag, so you don't need to import the table module with the long names, e.g. `newTable("loca")` or `newTable("glyf")`

RoelN replied:
👍 Thanks for the tip!

anthrotype commented on add_glyphs.py:
```diff
@@ -327,14 +342,30 @@ def add_ligature(lookup, cmap, seq, name):
   add_aliases_to_cmap(font, aliases)
   add_ligature_sequences(font, seqs, aliases)
-
+  if(add_cmap4):
```
```suggestion
  if add_cmap4:
```

anthrotype reviewed (approved):
LGTM, with minor comments

RoelN commented:
@anthrotype Thanks again for the pointers. All your suggestions have been implemented!

RoelN commented:
@anthrotype @rsheeter As okayed in the email thread, this can be merged. I don't have the rights to do that — could you do this? Thanks!

anthrotype commented:
Thank you, apologies for the delay

DeeDeeG commented:
A little late to suggest this, but I think the `glyf` table should probably only have one glyph for something obscure and unlikely to ever be used. Then it won't override real emoji characters in apps that don't support `CBLC/CBDT`.

But that implies some additional testing to make sure it works, and to confirm it's an actual improvement. I'm not 100% sure how to test, so I can't be totally confident this suggestion is right. Just wanted to mention the possibility. Maybe this sounds right to maintainers here?

Anyhow thanks for doing this PR!

RoelN commented:
@DeeDeeG I tried that by having only the `.notdef` present, but then the font wouldn't compile with ttx. Either the `glyf` table has to be omitted completely, or it should contain every single glyph. But I agree that'd be cleanest, if that'd work!

anthrotype commented:
note that having one glyph only or many identical glyphs doesn't make the glyph table shorter or longer. The loca will point to the same offset. The maxp table needs to contain the true number of glyphs, and be in agreement with all the tables in the font that contain glyphs (glyf or CBLC)

DeeDeeG commented:
Thanks for the replies.
